be/src/exec/dictionary_cache_writer.cpp

```diff
@@ -0,0 +1,334 @@
+    if (chunk != nullptr) {
+        if (chunk->num_rows() == 0) {
+            return Status::OK();
+        }
+
+        RETURN_IF_ERROR(DictionaryCacheWriter::ChunkUtil::check_chunk_has_null(*chunk.get()));
+
+        if (_buffer_chunk == nullptr) {
+            _buffer_chunk = chunk->clone_empty_with_slot();
+        }
+        DCHECK(_buffer_chunk != nullptr);
+        _buffer_chunk->append(*chunk.get());
+    }
+
+    if (_buffer_chunk != nullptr &&
+        (_buffer_chunk->container_memory_usage() > kChunkBufferLimit || chunk == nullptr /* end of appending*/)) {
+        DCHECK(_immutable_buffer_chunk == nullptr);
+        _immutable_buffer_chunk.swap(_buffer_chunk);
+    }
+
+    if (_immutable_buffer_chunk != nullptr) {
+        RETURN_IF_ERROR(_submit());
+        DCHECK(_buffer_chunk == nullptr && _immutable_buffer_chunk == nullptr);
+    }
+
+    if (terminate_flag != nullptr) {
+        terminate_flag->store(true, std::memory_order_release);
+    }
+
+    return Status::OK();
+}
+
+Status DictionaryCacheWriter::_submit() {
+    _num_pending_tasks.fetch_add(1, std::memory_order_release);
+    std::shared_ptr<Runnable> task(std::make_shared<RefreshDictionaryCacheTask>(this, _immutable_buffer_chunk));
+    auto st = ExecEnv::GetInstance()->dictionary_cache_pool()->submit(task);
+    if (!st.ok()) {
+        _num_pending_tasks.fetch_sub(1, std::memory_order_release);
+        DCHECK(_num_pending_tasks.load(std::memory_order_acquire) >= 0);
+        LOG(WARNING) << "refresh dictionary cache task submit failed, " << st.get_error_msg();
+    }
+    return st;
+}
+
+void DictionaryCacheWriter::sync_dictionary_cache(const Chunk* chunk) {
+    Status st = Status::OK();
+
+    // construct closures
+    std::vector<RefCountClosure<PRefreshDictionaryCacheResult>*> closures;
+    closures.resize(_t_dictionary_cache_sink.nodes.size());
+    for (auto& closure : closures) {
+        closure = new RefCountClosure<PRefreshDictionaryCacheResult>();
+        closure->ref();
+    }
+
+    DeferOp op([&]() {
+        _num_pending_tasks.fetch_sub(1, std::memory_order_release);
+        DCHECK(_num_pending_tasks.load(std::memory_order_acquire) >= 0);
+
+        for (auto& closure : closures) {
+            if (closure->unref()) {
+                delete closure;
+            }
+        }
+    });
+
+    auto schema = std::make_shared<OlapTableSchemaParam>();
+    (void)schema->init(_t_dictionary_cache_sink.schema);
+    auto pschema = std::make_unique<POlapTableSchemaParam>();
+    schema->to_protobuf(pschema.get());
+
+    while (true) {
+        std::unique_ptr<ChunkPB> pchunk = std::make_unique<ChunkPB>();
+        st = ChunkUtil::compress_and_serialize_chunk(chunk, pchunk.get());
+        if (!st.ok()) {
+            break;
+        }
+
+        st = _send_request(pchunk.get(), pschema.get(), closures);
+        if (!st.ok()) {
+            break;
+        }
+
+        st = _wait_response(closures);
+        break;
```

Comment:
always break?

Reply:
Yes, it always break. `While(True);break;`  is a way of implementing goto statements. It can do something. When there is an error in the middle of the code process, it can be jumped out in advance and error handling can be carried out directly